File: src/main/kotlin/com/foxpy/weather/WeatherGenerator.kt
package com.foxpy.weather

import java.time.LocalDate
import java.time.ZoneId
import java.util.Random


class WeatherGenerator {
    private val random = Random()
    private var currentWeather = ""

    private var lastUpdateDate = LocalDate.now(ZoneId.of("UTC")).minusDays(1)!!

    private val hotWeathers  = Array(20, {"rain"})    + Array(5,  {"thunderstorm"}) +
                               Array(5,  {"shower"})
    private val coldWeathers = Array(20, {"snow"})    + Array(5,  {"snowstorm"}) +
                               Array(5,  {"drizzle"})

    private val bothWeathers = Array(30, {"sunny"})   + Array(20, {"overcast"}) +
                               Array(15, {"cloudy"})  + Array(5,  {"fog"})

    private val windDirections = arrayListOf("➡️", "⬅️", "⬆️", "⬇️", "↗️", "↘️", "↙️", "↖️")

    fun getWeather(): String {
        if (lastUpdateDate != LocalDate.now(ZoneId.of("UTC"))) {
            updateDate()

            val temperature = getRandomNumber(-25, 35)
            val humidity = getRandomNumber(20, 100)
            val windSpeedMin = getRandomNumber(0, 10)
            val windSpeedMax = windSpeedMin + getRandomNumber(1, 10)
            val windDirection = windDirections[random.nextInt(windDirections.size)]
            val weathers = bothWeathers + if (temperature >= 0) hotWeathers else coldWeathers


            currentWeather = "Weather: ${weathers[random.nextInt(weathers.size)]}\n" +
                             "Temperature: $temperature" + "ºC\n" +
                             "Humidity: $humidity%\n" +
                             "Wind: $windDirection $windSpeedMin-$windSpeedMax m/s"
        }
        return currentWeather
    }

    fun changeWeather() {lastUpdateDate = LocalDate.now(ZoneId.of("UTC")).minusDays(1)}

    private fun getRandomNumber(min: Int, max: Int): Int {return random.nextInt(max + 1 - min) + min}
    private fun updateDate() {lastUpdateDate = LocalDate.now(ZoneId.of("UTC"))}
}

File: weathergenerator.py
from random import choice, randint

degrees = chr(0xBA) + "C"
wind_directions = [chr(0x2B06) + chr(0xFE0F),   # N
                   chr(0x27A1) + chr(0xFE0F),   # E
                   chr(0x2B05) + chr(0xFE0F),   # W
                   chr(0x2B07) + chr(0xFE0F),   # S
                   chr(0x2197) + chr(0xFE0F),   # NE
                   chr(0x2196) + chr(0xFE0F),   # NW
                   chr(0x2198) + chr(0xFE0F),   # SE
                   chr(0x2199) + chr(0xFE0F)]   # SW


class WeatherType:
    def __init__(self, weathers, temperature_day_range,
            temperature_night_offset_range, wind_range, humidity_range):
        self.__weathers = weathers
        self.__min_temp_day, self.__max_temp_day = \
                tuple(temperature_day_range)
        self.__min_temp_night_offset, self.__max_temp_night_offset = \
                tuple(temperature_night_offset_range)
        self.__min_wind_speed, self.__max_wind_speed, \
        self.__min_wind_speed_offset, self.__max_wind_speed_offset = \
                tuple(wind_range)
        self.__min_humidity, self.__max_humidity = \
                tuple(humidity_range)

    def get_weather(self):
        weather = choice(self.__weathers)
        temperature_day = randint(self.__min_temp_day, self.__max_temp_day)
        temperature_night = temperature_day - \
                randint(self.__min_temp_night_offset,
                        self.__max_temp_night_offset)
        humidity = randint(self.__min_humidity, self.__max_humidity)
        wind_speed_min = randint(self.__min_wind_speed, self.__max_wind_speed)
        wind_speed_max = wind_speed_min + \
                randint(self.__min_wind_speed_offset,
                        self.__max_wind_speed_offset)
        wind_direction = choice(wind_directions)

        return f"Weather: {weather}\n" + \
            f"Temperature day: {temperature_day}{degrees}\n" + \
            f"Temperature night: {temperature_night}{degrees}\n" + \
            f"Humidity: {humidity}%\n" + \
            f"Wind speed: {wind_direction} {wind_speed_min}-{wind_speed_max}" +\
            " m/s\n"


class WeatherGenerator():
    weather_types = \
            [WeatherType(["sunny", "cloudy", "rain"],
                    (3, 40), (2, 8), (2, 10, 3, 10), (10, 90)),
            WeatherType(["sunny", "cloudy", "snow"],
                    (-25, 0), (3, 10), (2, 10, 3, 10), (10, 90)),
            WeatherType(["fog", "overcast", "drizzle"],
                    (2, 22), (2, 4), (0, 3, 1, 2), (80, 95)),
            WeatherType(["thunderstorm"],
                    (3, 30), (2, 6), (4, 14, 5, 12), (60, 95))]

    def get_weather(self):
        return choice(self.weather_types).get_weather()
